autoload/ale/code_action.vim

```diff
@@ -127,10 +127,22 @@ function! ale#code_action#ApplyChanges(filename, changes, should_save) abort
         endif
 
         call extend(l:middle, l:insertions[1:])
-        let l:middle[-1] .= l:lines[l:end_line - 1][l:end_column - 1 :]
+
+        if l:end_line <= len(l:lines)
+            " Only extend the last line if end_line is within the range of
+            " lines.
+            let l:middle[-1] .= l:lines[l:end_line - 1][l:end_column - 1 :]
+        endif
 
         let l:lines_before_change = len(l:lines)
-        let l:lines = l:start + l:middle + l:lines[l:end_line :]
+
+        if l:end_line < len(l:lines)
+            let l:end = l:lines[l:end_line :]
+        else
+            let l:end = []
+        endif
```

Comment:
You don't need to check the length for a slice. You'll only ever get an error when you can't access an index. `x[1]` might raise an error, but `x[1:]` never will, it will give you `[]`.